pkgs/java_http/lib/src/java_client.dart

```diff
@@ -39,51 +41,98 @@ class JavaClient extends BaseClient {
     // See https://github.com/dart-lang/http/pull/980#discussion_r1253700470.
     _initJVM();
 
+    final receivePort = ReceivePort();
+    final events = StreamQueue<dynamic>(receivePort);
+
     // We can't send a StreamedRequest to another Isolate.
     // But we can send Map<String, String>, String, UInt8List, Uri.
-    final requestBody = await request.finalize().toBytes();
-    final requestHeaders = request.headers;
-    final requestMethod = request.method;
-    final requestUrl = request.url;
-
-    final (statusCode, reasonPhrase, responseHeaders, responseBody) =
-        await Isolate.run(() async {
-      final httpUrlConnection = URL
-          .ctor3(requestUrl.toString().toJString())
-          .openConnection()
-          .castTo(HttpURLConnection.type, deleteOriginal: true);
-
-      requestHeaders.forEach((headerName, headerValue) {
-        httpUrlConnection.setRequestProperty(
-            headerName.toJString(), headerValue.toJString());
-      });
-
-      httpUrlConnection.setRequestMethod(requestMethod.toJString());
-      _setRequestBody(httpUrlConnection, requestBody);
-
-      final statusCode = _statusCode(requestUrl, httpUrlConnection);
-      final reasonPhrase = _reasonPhrase(httpUrlConnection);
-      final responseHeaders = _responseHeaders(httpUrlConnection);
-      final responseBody = _responseBody(httpUrlConnection);
-
-      httpUrlConnection.disconnect();
-
-      return (
-        statusCode,
-        reasonPhrase,
-        responseHeaders,
-        responseBody,
-      );
-    });
-
-    return StreamedResponse(Stream.value(responseBody), statusCode,
-        contentLength:
-            _contentLengthHeader(request, responseHeaders, responseBody.length),
+    final isolateRequest = (
+      sendPort: receivePort.sendPort,
+      url: request.url,
+      method: request.method,
+      headers: request.headers,
+      body: await request.finalize().toBytes(),
+    );
+
+    // Could create a new class to hold the data for the isolate instead
+    // of using a record.
+    final httpRequestIsolate = await Isolate.spawn(
+      _isolateMethod,
+      isolateRequest,
+    );
+    //httpRequestIsolate.errors.listen(print);
+
+    final statusCode = await events.next as int;
+    final reasonPhrase = await events.next as String?;
+    final responseHeaders = await events.next as Map<String, String>;
+    final responseBody = events.rest
```

Comment:
You should probably either listen to `events.rest` or use a [`StreamTransformer`](https://api.flutter.dev/flutter/dart-async/Stream/Stream.eventTransformed.html)

And then dispatch based on type e.g.

```dart
if (object is ClientException) {
   sink.addError(object)
} else if (object is List) {
  sink.add(object);
}
```

Reply:
I ended up listening to `events.rest` as using a `StreamTransformer` looked a bit fiddly. Perhaps in the future, I could refactor to use a `StreamTransformer` to make the code cleaner.

Thanks for your help!